File: masonry/src/engine/model-version.js
'use strict';
// engine/model-version.js — Content hash of simulate.py + constants.py for finding correlation.
//
// Port of bl/model_version.py to Node.js.

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

function computeModelHash(projectRoot) {
  const root = String(projectRoot);
  let content = Buffer.alloc(0);
  let found = false;

  for (const fname of ['simulate.py', 'constants.py']) {
    const fpath = path.join(root, fname);
    try {
      const data = fs.readFileSync(fpath);
      content = Buffer.concat([content, data]);
      found = true;
    } catch {
      // file doesn't exist — skip
    }
  }

  if (!found) return 'no-model';
  return crypto.createHash('sha256').update(content).digest('hex').slice(0, 12);
}

function embedInFinding(content, modelHash) {
  if (content.includes('**Model hash**:')) return content;
  return content.trimEnd() + `\n\n**Model hash**: ${modelHash}\n`;
}

module.exports = { computeModelHash, embedInFinding };

File: bl/model_version.py
"""Compute a content hash of simulate.py + constants.py for finding correlation."""
import hashlib
from pathlib import Path


def compute_model_hash(project_root) -> str:
    """SHA-256 of simulate.py + constants.py content, first 12 hex chars.
    Returns 'no-model' if neither file exists."""
    root = Path(project_root)
    content = b""
    found = False
    for f in [root / "simulate.py", root / "constants.py"]:
        if f.exists():
            content += f.read_bytes()
            found = True
    return "no-model" if not found else hashlib.sha256(content).hexdigest()[:12]


def embed_in_finding(content: str, model_hash: str) -> str:
    """Append '**Model hash**: {hash}' to finding content if not already present."""
    if "**Model hash**:" in content:
        return content
    return content.rstrip() + f"\n\n**Model hash**: {model_hash}\n"
